### PR TITLE
Feature/agnews dataset

### PR DESCRIPTION
**Pull request by matprst**
Add the AGNews dataset (both text and graph) to the workflow.

### CODE REVIEW

File: data_prep/agnews_graph.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        train_docs, test_docs, val_docs = docs
+        train_labels, test_labels, val_labels = labels
+
+        all_docs = train_docs + test_docs + val_docs
+        all_labels = train_labels + test_labels + val_labels
+
+        tokenizer = WordPunctTokenizer() # TODO: look into a better one?
+        corpus = [tokenizer.tokenize(sentence) for sentence in all_docs]
+
+        print('Compute tf.idf')
```
Comment (rahelbeloch) [inline data_prep/agnews_graph.py:37]: Apart from this, we kind of have the exact same code as in `ReutersGraph`, except some lines, e.g. line 55. Can we somehow reduce the amount of duplicate code? Maybe e.g. creating a superclass `GraphDataset` from which `AGnewsGraph` and `ReutersGraph` inherit which does all the common stuff for graph datasets and then for specific stuff like the preprocessing calls the respective methods which are overwritten by the respective graph specific classes.

File: data_prep/agnews_graph.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        node_feats = torch.eye(len(self.iton), device=self.device).float()
+        print('Features mtx is {} GBs in size'.format(node_feats.nelement() * node_feats.element_size() * 1e-9))
+
+        # Create pytorch geometric format data
+        self.data = Data(x=node_feats, edge_index=edge_index, edge_attr=edge_attr, y=ntol)
+        self.data.train_mask = train_mask
+        self.data.val_mask = val_mask
+        self.data.test_mask = test_mask
+
+    def generate_edges(self, num_docs, tf_idf, pmi_scores):
```
Comment (rahelbeloch) [inline data_prep/agnews_graph.py]: I think this whole method is duplicate with the method in `ReutersGraph`?

File: data_prep/agnews_graph.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            edge_index.append([w_i_ind, w_j_ind])
+            edge_index.append([w_j_ind, w_i_ind])
+            edge_attr.append(score)
+            edge_attr.append(score)
+
+        edge_index = torch.tensor(edge_index).t().contiguous()
+        edge_attr = torch.tensor(edge_attr).float()
+        return edge_index, edge_attr
+
+    def generate_masks(self, train_num, val_num, test_num):
```
Comment (rahelbeloch) [inline data_prep/agnews_graph.py]: Same as above

File: data_prep/agnews_graph.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        # extract the text for each news article        
+        train_texts, train_labels = zip(*[(data["text"], data["label"]) for data in train_val_splits["train"]])
+        val_texts, val_labels = zip(*[(data["text"], data["label"]) for data in train_val_splits["test"]])
+        test_texts, test_labels = zip(*[(data["text"], data["label"]) for data in dataset["test"]])
+
+        unique_classes = train_val_splits["train"].features["label"].names
+
+         # For testing with only a few docs:
+        docs = (list(train_texts[:1000]), list(test_texts[:100]), list(val_texts[:100]))
+        labels = (list(train_labels[:1000]), list(test_labels[:100]), list(val_labels[:100]))
```
Comment (rahelbeloch) [inline data_prep/agnews_graph.py]: Would maybe be nice if these number of docs could be passed to the class / method, so that whoever uses this Dataset class can simply pass the number of docs which should be used.

File: agnews_graph.ipynb [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+   "id": "5373b31a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import torch\n",
+    "import numpy as np\n",
+    "\n",
+    "from data_prep.agnews_graph import AGNewsGraph\n",
+    "\n",
+    "device = torch.device(\"cuda:0\") if torch.cuda.is_available() else torch.cuda(\"cpu\")"
```
Comment (rahelbeloch) [inline agnews_graph.ipynb]: I think it should be `torch.device(\"cpu\")` in the end

File: data_prep/agnews_text.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        unique_cls = dataset.features["label"].names
+        return cls(texts, labels, unique_cls, tokenizer)
+    
+    @staticmethod
+    def _prepare_split(dataset):
+        texts = []
+        labels = []
+        for data in dataset:
+            texts.append(data["text"])
+            labels.append(data["label"])
```
Comment (rahelbeloch) [inline data_prep/agnews_text.py]: Could do this with two list comprehensions in two lines

File: agnews_text.ipynb [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "10ddc58e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from torch.utils.data import DataLoader\n",
+    "from transformers import RobertaTokenizer\n",
+    "from data_prep.agnews_text import AGNews"
```
Comment (rahelbeloch) [inline agnews_text.ipynb]: Should be `import AGNewsText`?

File: data_prep/agnews_text.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        return len(self.labels)
+
+
+if __name__ == "__main__":
+    from transformers import RobertaTokenizer
+    from torch.utils.data import DataLoader
+
+    tokenizer = RobertaTokenizer.from_pretrained("roberta-base")
+
+    train_set, test_set, val_set = AGNews.splits(tokenizer)
```
Comment (rahelbeloch) [inline data_prep/agnews_text.py:83]: I guess it should be `AGNewsText` here

File: data_prep/agnews_text.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    print("test size=", len(test_set))
+    print("val size=", len(val_set))
+
+    train_loader = DataLoader(train_set, batch_size=2, shuffle=True, collate_fn=train_set.get_collate_fn())
+
+    for data_batch in train_loader:
+        print(data_batch)
+        input_ids = data_batch["input_ids"]
+        attention_mask = data_batch["attention_mask"]
+        d_labels = data_batch["labels"]
```
Comment (rahelbeloch) [inline data_prep/agnews_text.py:94]: Currently, `AGNewsText` only has keys `text` and `label`

Comment (rahelbeloch) [reply data_prep/agnews_text.py:94]: Might rokenization for roberta still be missing?

File: agnews_text.ipynb [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+     ]
+    }
+   ],
+   "source": [
+    "agnews_train_dataloader = DataLoader(agnews_train, batch_size=2, shuffle=True)\n",
+    "\n",
+    "for data in agnews_train_dataloader:\n",
+    "    input_ids = data[\"input_ids\"]\n",
+    "    attention_mask = data[\"attention_mask\"]\n",
+    "    labels = data[\"labels\"]\n",
```
Comment (rahelbeloch) [inline agnews_text.ipynb:120]: Currently, `AGNewsText` only has keys `text` and `label` - toke nization for roberta missing?

File: train.py [hunk 11]
```diff
@@ -146,6 +147,8 @@ def get_dataset(dataset_name):
         return R8Text
     elif dataset_name == "R52Text":
         return R52Text
+    elif dataset_name == "AGNews":
```
Comment (rahelbeloch) [inline train.py]: Maybe better use `AGNewsText` as R52 is also `R52Text`

**Review by rahelbeloch — CHANGES_REQUESTED** [review]

**Comment by rahelbeloch** [comment]
@MatPrst Reviewed... Some of the things I already mentioned during our call today.

File: agnews_text.ipynb (hunk 10 above)
Comment (matprst) [reply agnews_text.ipynb:120]: Good spot, I forgot to pass the collate function to the Dataloader

**Review by rahelbeloch — APPROVED** [review]
@MatPrst Looks good to me, you can merge it from my side
